### PR TITLE
[pipe-push-pop] Refactor to add Drop

### LINKED ISSUE #720: [tests] Refactor `pipe-push-pop` to Follow "Associated Functions" Design
## Context

The `pipe-push-pop` application was written as a collection of standalone functions. 

We should refactor the `pipe-push-pop` application to follow an "associated functions" code structure, so that we can have correct trait implementations, such as `Drop`.

### PR DESCRIPTION
**Pull request by anandbonde**
Closes #720

### CODE REVIEW

File: examples/rust/pipe-push-pop.rs [hunk 1]
```diff
@@ -100,130 +98,135 @@ fn push_and_wait(libos: &mut LibOS, pipeqd: QDesc, length: usize, value: u8) ->
...
+                anyhow::bail!("expected a valid sga");
             }
-        }
 
-        if libos.sgafree(sga).is_err() {
-            // If error, close pipes.
-            close(&mut libos, pipeqd);
-            anyhow::bail!("failed to release scatter-gather array");
-        };
+            println!("pop - {:?} bytes", nbytes);
```
Comment (ppenna) [inline examples/rust/pipe-push-pop.rs:169]: Could you remove the '-'?

**Review by ppenna — APPROVED** [review]
LGTM. I added some minor comments, though.

File: examples/rust/pipe-push-pop.rs (hunk 1 above)
Comment (anandbonde) [reply examples/rust/pipe-push-pop.rs]: Done.

File: examples/rust/pipe-push-pop.rs [hunk 2]
```diff
@@ -100,130 +98,135 @@ fn push_and_wait(libos: &mut LibOS, pipeqd: QDesc, length: usize, value: u8) ->
...
+            if let Some(sga) = self.sga {
+                // Sanity received data.
+                let recvbuf: &[u8] = unsafe {
+                    slice::from_raw_parts(
+                        sga.sga_segs[0].sgaseg_buf as *const u8,
+                        sga.sga_segs[0].sgaseg_len as usize,
+                    )
+                };
+                for x in &recvbuf[..] {
+                    assert_eq!(*x, round);
```
[the remark this answers is not in the document]
Comment (anandbonde) [reply examples/rust/pipe-push-pop.rs]: Done.

**Pull request merged**